src/base/BaseActionsRouter.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity ^0.8.24;
+
+import {IPoolManager} from "@uniswap/v4-core/src/interfaces/IPoolManager.sol";
+import {Currency} from "@uniswap/v4-core/src/types/Currency.sol";
+import {TransientStateLibrary} from "@uniswap/v4-core/src/libraries/TransientStateLibrary.sol";
+
+import {SafeTransferLib} from "solmate/utils/SafeTransferLib.sol";
+
+import {SafeCallback} from "./SafeCallback.sol";
+import {BytesLib} from "../libraries/BytesLib.sol";
+
+/// @notice Abstract contract for performing a combination of actions on Uniswap v4.
+/// @dev Suggested uint256 action values are defined in Actions.sol, however any definition can be used
+abstract contract BaseActionsRouter is SafeCallback {
+    using TransientStateLibrary for IPoolManager;
+    using SafeTransferLib for address;
+    using BytesLib for bytes;
+
+    /// @notice emitted when different numbers of parameters and actions are provided
+    error LengthMismatch();
+
+    /// @notice emitted when an inheriting contract does not support an action
+    error UnsupportedAction(uint256 action);
+
+    constructor(IPoolManager poolManager) SafeCallback(poolManager) {}
+
+    /// @notice internal function that triggers the execution of a set of actions on v4
+    /// @dev inheriting contracts should call this function to trigger execution
+    function _executeActions(bytes calldata params) internal {
+        poolManager.unlock(params);
+    }
+
+    /// @notice function that is called by the PoolManager through the SafeCallback.unlockCallback
+    function _unlockCallback(bytes calldata data) internal override returns (bytes memory) {
```

Comment:
i think this line introduces a compiler warning - I think @gretzke had a way to fix but not sure? Could just return empty bytes